HQ Hunter fix

hessenfarmer opened the pull request:
fixes #4681
adds documentation to start conditions

hessenfarmer committed:
add a luahook to get all constructionsites of a player

hessenfarmer committed:
add HQ hunter win condition

hessenfarmer committed:
made a defeated player loose his constructionsites, military sites and ships

hessenfarmer committed:
polished the defeated check

hessenfarmer committed:
Added a real "no Rule" game. adopted the legacy endless WC removing no rule as players defeated are sorted out there

hessenfarmer committed:
added a not that HQ Hunter does not work with Poor Hamlet, Fortified Village, Discovery and New World Sc

hessenfarmer committed:
fixed a formatting issue

Commit:
'data/scripting/win_conditions/hq_hunter.lua' was automatically formatted.

hessenfarmer committed:
Update src/scripting/lua_game.cc

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
removed no rules
made endless game ignore player defeated

hessenfarmer committed:
pulled out common code as requested in review

hessenfarmer committed:
corrected errors

Commit:
'src/scripting/lua_game.cc' was automatically formatted.

hessenfarmer committed:
try to make codechaeck happy

hessenfarmer committed:
Merge branch '3849_win_conditions' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
Update src/scripting/lua_game.cc

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
implemented review results

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
readded erroneausly deleted include

hessenfarmer committed:
Update endless_game_fogless.lua

hessenfarmer committed:
get rid of hardcoding in HQ hunter

hessenfarmer committed:
improved loops

hessenfarmer committed:
fixed HQ Hunter

hessenfarmer committed:
fixed HQ hunter for discovery new world

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
introduced a safeguard to avoid map object not available error

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 3849_win_conditions

frankystone committed:
Merge branch 'master' into 3849_win_conditions

hessenfarmer committed:
introduce incompatible winconditions check

hessenfarmer committed:
next attempt

hessenfarmer committed:
added the MP part hopefully

Commit:
2 files were automatically formatted.

hessenfarmer committed:
fixed whitespace

hessenfarmer committed:
Merge branch '3849_win_conditions' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
replaced HQ Hunter with the filename of the wc

hessenfarmer committed:
update the startconditions if wincondition is changed

hessenfarmer committed:
made it work for singleplayer

hessenfarmer committed:
Included Arne Schauf to coders as he had the idea and design of this in the first place
Took the freedom to add myself to coders as well

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
adjusted HQ hunter and autocrat as required

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 3849_win_conditions

frankystone committed:
Merge branch 'master' into 3849_win_conditions

hessenfarmer committed:
Merge branch 'master' into 3849_win_conditions

hessenfarmer committed:
fix typo

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
streamline compatible win condition script

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
Merge branch 'master' into 3849_win_conditions

hessenfarmer committed:
Merge branch '3849_win_conditions' of https://github.com/widelands/widelands into 3849_win_conditions

Commit:
'src/ui_fsmenu/singleplayerdropdown.cc' was automatically formatted.

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
Merge branch '3849_win_conditions' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
implemented suggestion

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
Merge branch '3849_win_conditions' of https://github.com/widelands/widelands into 3849_win_conditions

Commit:
'src/ui_fsmenu/multiplayersetupgroup.cc' was automatically formatted.

hessenfarmer committed:
Merge branch '3849_win_conditions' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
added TODO to pull out common code

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 3849_win_conditions

hessenfarmer committed:
destroy also conquered milsites

hessenfarmer committed:
added documentation for start conditions

Commit:
'data/tribes/initialization/atlanteans/starting_conditions/discovery.lua' was automatically formatted.

Noordfrees commented:
We use the terms "Start condition" and "Starting condition" interchangeably currently. This should be unified. Which term do we want? I think "Starting condition" is used in more places than "Start condition".

Noordfrees commented:
I cannot find a link to the new documentation from the HTML file tree. I think this needs a cross-reference from autogen_toc_lua_tribes_defining or such and an entry in some TOC.
```
checking consistency... /home/benedikt/wl/widelands/doc/sphinx/source/autogen_auxiliary_mapobjects.rst: WARNING: document isn't included in any toctree
/home/benedikt/wl/widelands/doc/sphinx/source/autogen_lua_tribes_defining_atlanteans.rst: WARNING: document isn't included in any toctree
/home/benedikt/wl/widelands/doc/sphinx/source/autogen_lua_tribes_defining_preload.rst: WARNING: document isn't included in any toctree
/home/benedikt/wl/widelands/doc/sphinx/source/autogen_lua_tribes_other_global_helptexts.rst: WARNING: document isn't included in any toctree
```

hessenfarmer committed:
fix typo

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
another typo

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
clarification about map tags

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
one more typo

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
typo again

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer commented:
> 
> 
> I cannot find a link to the new documentation from the HTML file tree. I think this needs a cross-reference from autogen_toc_lua_tribes_defining or such and an entry in some TOC.
> 
> ```
> checking consistency... /home/benedikt/wl/widelands/doc/sphinx/source/autogen_auxiliary_mapobjects.rst: WARNING: document isn't included in any toctree
> /home/benedikt/wl/widelands/doc/sphinx/source/autogen_lua_tribes_defining_atlanteans.rst: WARNING: document isn't included in any toctree
> /home/benedikt/wl/widelands/doc/sphinx/source/autogen_lua_tribes_defining_preload.rst: WARNING: document isn't included in any toctree
> /home/benedikt/wl/widelands/doc/sphinx/source/autogen_lua_tribes_other_global_helptexts.rst: WARNING: document isn't included in any toctree
> ```

Well I hoped to achieve this with the entry in extract_rst.py as the TOC is autogenerated from this as far as I understand it.
Will try to setup the doc building in my MSYS environment to investigate

frankystone commented:
Please add also a link in [Add-ons starting condition](https://www.widelands.org/documentation/add-ons/#starting-condition) which links to your new documentation.

hessenfarmer committed:
fixed doc Generation

hessenfarmer committed:
Try to address review comment

Commit:
'data/tribes/initialization/atlanteans/starting_conditions/discovery.lua' was automatically formatted.

hessenfarmer committed:
next try

Commit:
'data/tribes/initialization/atlanteans/starting_conditions/discovery.lua' was automatically formatted.

hessenfarmer committed:
indentation matters obviously

hessenfarmer committed:
Included link to start conditions in Add on documentation

frankystone commented:
If you want to address unifying of 'start condition' vs. 'starting condition' in this branch here is a list.

The number after the filename is the line number:
~~~~
git branch: (3849_win_conditions)$:> grep -rni 'start condition' data/
data/tribes/initialization/frisians/starting_conditions/headquarters.lua:2:--                 Start conditions for Headquarters
data/tribes/initialization/frisians/starting_conditions/discovery.lua:2:--                 Start conditions for Discovery
data/tribes/initialization/frisians/starting_conditions/new_world.lua:2:--                 Start conditions for New World
data/tribes/initialization/amazons/starting_conditions/headquarters.lua:2:--                 Start conditions for Headquarters
data/tribes/initialization/empire/starting_conditions/discovery.lua:2:--                 Start conditions for Discovery
data/tribes/initialization/empire/starting_conditions/new_world.lua:2:--                 Start conditions for New World
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:4:-- Start Conditions
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:7:-- The start conditions a tribe can use are listed in the tribes :ref:`init.lua <lua_tribes_tribes_init>` file.
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:14:--    This function initializes the start condition.
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:16:--    :arg table: This table contains all the data needed to initialize the start condition.
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:19:--    **descname**: The localized name of the start condition.
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:21:--    **tooltip**: A short description of the start condition. Should contain essential hints if necessary.
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:24:--                  e.g. ``{ "seafaring" }`` if the start conditions needs a seafaring map.
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:25:--                  If present, the start condition can only be selected by the user if the map contains *all* listed tags.
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:28:--                         start condition, e.g. ``{ "scripting/win_conditions/hq_hunter.lua" }`` if the start condition
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:31:--    **func**: A standardized function to determine whether to share this start condition between players (Shared Kingdom)
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:34:--              together with the start condition.
data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:111:--                 Start conditions for Discovery
data/tribes/initialization/atlanteans/starting_conditions/new_world.lua:2:--                 Start conditions for New World
data/tribes/initialization/barbarians/starting_conditions/headquarters.lua:2:--                 Start conditions for Headquarters
data/tribes/initialization/barbarians/starting_conditions/discovery.lua:2:--                 Start conditions for Discovery
data/tribes/initialization/barbarians/starting_conditions/fortified_village.lua:2:--                    Start conditions for Fortified Village
data/tribes/initialization/barbarians/starting_conditions/village.lua:2:--                    Start conditions for Village
data/tribes/initialization/barbarians/starting_conditions/new_world.lua:2:--                 Start conditions for New World
~~~~

hessenfarmer commented:
> 
> 
> If you want to address unifying of 'start condition' vs. 'starting condition' in this branch here is a list.
> 
> The number after the filename is the line number:
> 
> ```
> git branch: (3849_win_conditions)$:> grep -rni 'start condition' data/
> data/tribes/initialization/frisians/starting_conditions/headquarters.lua:2:--                 Start conditions for Headquarters
> data/tribes/initialization/frisians/starting_conditions/discovery.lua:2:--                 Start conditions for Discovery
> data/tribes/initialization/frisians/starting_conditions/new_world.lua:2:--                 Start conditions for New World
> data/tribes/initialization/amazons/starting_conditions/headquarters.lua:2:--                 Start conditions for Headquarters
> data/tribes/initialization/empire/starting_conditions/discovery.lua:2:--                 Start conditions for Discovery
> data/tribes/initialization/empire/starting_conditions/new_world.lua:2:--                 Start conditions for New World
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:4:-- Start Conditions
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:7:-- The start conditions a tribe can use are listed in the tribes :ref:`init.lua <lua_tribes_tribes_init>` file.
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:14:--    This function initializes the start condition.
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:16:--    :arg table: This table contains all the data needed to initialize the start condition.
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:19:--    **descname**: The localized name of the start condition.
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:21:--    **tooltip**: A short description of the start condition. Should contain essential hints if necessary.
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:24:--                  e.g. ``{ "seafaring" }`` if the start conditions needs a seafaring map.
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:25:--                  If present, the start condition can only be selected by the user if the map contains *all* listed tags.
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:28:--                         start condition, e.g. ``{ "scripting/win_conditions/hq_hunter.lua" }`` if the start condition
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:31:--    **func**: A standardized function to determine whether to share this start condition between players (Shared Kingdom)
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:34:--              together with the start condition.
> data/tribes/initialization/atlanteans/starting_conditions/discovery.lua:111:--                 Start conditions for Discovery
> data/tribes/initialization/atlanteans/starting_conditions/new_world.lua:2:--                 Start conditions for New World
> data/tribes/initialization/barbarians/starting_conditions/headquarters.lua:2:--                 Start conditions for Headquarters
> data/tribes/initialization/barbarians/starting_conditions/discovery.lua:2:--                 Start conditions for Discovery
> data/tribes/initialization/barbarians/starting_conditions/fortified_village.lua:2:--                    Start conditions for Fortified Village
> data/tribes/initialization/barbarians/starting_conditions/village.lua:2:--                    Start conditions for Village
> data/tribes/initialization/barbarians/starting_conditions/new_world.lua:2:--                 Start conditions for New World
> ```

Many thanks. That helps a lot. and yes I definitly want to include the unified term in this branch. 
Another question is there any way to test the documentation changes in a website test?

frankystone commented:
> Another question is there any way to test the documentation changes in a website test?

I did that already :-)

frankystone commented:
Here are the zipped .html files including the .css made with your branch and the website code.
[current.zip](https://github.com/widelands/widelands/files/5988335/current.zip)

Just unpack it somewhere and open the file `path\current\index.html`

hessenfarmer committed:
harmonized usage of starting condition
Harmonized init function of sc to comply with documentation
fixed a string issue with amazons poor hamlet

hessenfarmer committed:
deleted superfluous entry

hessenfarmer committed:
clarified folder name

hessenfarmer committed:
fix docu format

hessenfarmer committed:
fix a link

Noordfrees reviewed (commented):
A few more nits, then this is good to go.

Code for the HQ hunter bug LGTM, not tested.

hessenfarmer committed:
implemented suggestion about linking documentation

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
fixed a last instance of not harmonized string

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
typo fix

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
typo fix

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
typo again

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
last typo

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
missing point

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
implemented suggestion

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

hessenfarmer committed:
Merge branch 'master' into 3849_win_conditions

Noordfrees committed:
linkfix

hessenfarmer commented:
@frankystone have you been able to test whether the linked bug gets fixed with this PR?

frankystone commented:
No, sorry.

hessenfarmer commented:
Tested it myself. Works as expected. Here is the replay


[replays.zip](https://github.com/widelands/widelands/files/5991409/replays.zip)
